In [7]:
from tensorflow import keras
import numpy as np
from tensorflow.keras import layers, models
from tensorflow.keras.datasets import mnist

In [5]:
inputs = keras.Input(shape=(28,28,1))
x = layers.Conv2D(filters=32, kernel_size=3, activation='relu')(inputs)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=64, kernel_size=3, activation='relu')(x)
x = layers.MaxPooling2D(pool_size=2)(x)
x = layers.Conv2D(filters=128, kernel_size=3, activation='relu')(x)
x = layers.Flatten()(x)
output = layers.Dense(10, activation='softmax')(x)  # classify the minist images

c_model = models.Model(inputs=inputs, outputs=output)

In [6]:
c_model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_2 (InputLayer)        [(None, 28, 28, 1)]       0         
                                                                 
 conv2d (Conv2D)             (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d (MaxPooling2  (None, 13, 13, 32)        0         
 D)                                                              
                                                                 
 conv2d_1 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_1 (MaxPoolin  (None, 5, 5, 64)          0         
 g2D)                                                            
                                                                 
 conv2d_2 (Conv2D)           (None, 3, 3, 128)         73856 

### Effect of Covnet on input 
* The inputs are in (image_height, image_width, channels) format
* As we can see as soon as we put a conv2d network on the input the height and width decrease in size by 2 and the channels increase. 
* Max pool net halfs the height and width keeping the channels the same 
  
At the end we have final covnet reduce the set to (3,3,128) which is then flattened to 1152 and finally a softmax is applied with a Dense layer. 

Next we train the model with the minist dataset

In [8]:

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()
train_images = train_images.reshape((60000, 28, 28, 1)).astype('float32') / 255
test_images = test_images.reshape((10000, 28, 28, 1)).astype('float32') / 255

c_model.compile(optimizer='rmsprop',
                loss='sparse_categorical_crossentropy',
                metrics=['accuracy'])
c_model.fit(train_images, train_labels, epochs=5, batch_size=64)


Epoch 1/5
938/938 [==============================] - 12s 12ms/step - loss: 0.1569 - accuracy: 0.9509
Epoch 2/5
938/938 [==============================] - 12s 13ms/step - loss: 0.0444 - accuracy: 0.9865
Epoch 3/5
938/938 [==============================] - 11s 12ms/step - loss: 0.0305 - accuracy: 0.9906
Epoch 4/5
938/938 [==============================] - 11s 12ms/step - loss: 0.0236 - accuracy: 0.9928
Epoch 5/5
938/938 [==============================] - 11s 12ms/step - loss: 0.0186 - accuracy: 0.9940


In [9]:
# next we will use the model to get accuracy. 
test_loss, test_acc = c_model.evaluate(test_images, test_labels)
print(f"Test Accuracy: {test_acc:.3f}")

313/313 [==============================] - 1s 3ms/step - loss: 0.0227 - accuracy: 0.9927
Test Accuracy: 0.993


**Accuracy** 
The accuracy with the convnet model is 99.3 % compared with the 97.1% with the dense layers we calculated earlier in chapter 5. 